packages/babel-helper-modules/src/normalize-and-load-metadata.js

```diff
@@ -0,0 +1,378 @@
+  reexports: Map<string, string>,
+
+  // List of names to re-export namespace as.
+  reexportNamespace: Set<string>,
+
+  // Tracks if the source should be re-exported.
+  reexportAll: null | {
+    loc: ?Node,
+  },
+};
+
+export type LocalExportMetadata = {
+  name: Array<string>, // names of exports
+  kind: "import" | "hoisted" | "block" | "var",
+};
+
+/**
+ * Check if the module has any exports that need handling.
+ */
+export function hasExports(metadata: ModuleMetadata) {
+  const { local, source } = metadata;
+
+  return (
+    local.size > 0 ||
+    Array.from(source).some(([, meta]) => {
+      return (
+        meta.reexports.size > 0 ||
+        meta.reexportNamespace.size > 0 ||
+        !!meta.reexportAll
+      );
+    })
+  );
+}
+
+/**
+ * Check if a given source is an anonymous import, e.g. "import 'foo';"
+ */
+export function isAnonymousImport(source: SourceModuleMetadata) {
+  return (
+    source.imports.size === 0 &&
+    source.importsNamespace.size === 0 &&
+    source.reexports.size === 0 &&
+    source.reexportNamespace.size === 0 &&
+    !source.reexportAll
+  );
+}
+
+/**
+ * Remove all imports and exports from the file, and return all metadata
+ * needed to reconstruct the module's behavior.
+ */
+export default function normalizeModuleAndLoadMetadata(
+  programPath: NodePath,
+  exportName?: string,
+  { strict = false, noInterop = false } = {},
+): ModuleMetadata {
+  if (!exportName) {
+    exportName = programPath.scope.generateUidIdentifier("exports").name;
+  }
+
+  const localData = normalizeLocalExports(programPath);
+
+  const sourceData = new Map();
+  const getData = sourceNode => {
+    const source = sourceNode.value;
+
+    let data = sourceData.get(source);
+    if (!data) {
+      data = {
+        // The "StringLiteral" AST node from the import/export statement.
+        sourceNode,
+        statements: [],
+
+        name: programPath.scope.generateUidIdentifier(
+          basename(source, extname(source)),
+        ).name,
+
+        interop: "none",
+
+        // Data about the requested sources and names.
+        imports: new Map(),
+        importsNamespace: new Set(),
+
+        // Metadata about data that is passed directly from source to export.
+        reexports: new Map(),
+        reexportNamespace: new Set(),
+        reexportAll: null,
+      };
+      sourceData.set(source, data);
+    }
+    return data;
+  };
+
+  programPath.get("body").forEach(child => {
+    if (child.isImportDeclaration()) {
+      const data = getData(child.node.source);
+      data.statements.push(child.node);
+
+      child.get("specifiers").forEach(spec => {
+        if (spec.isImportDefaultSpecifier()) {
+          if (data.interop === "none") data.interop = "default";
+
+          const localName = spec.get("local").node.name;
+
+          data.imports.set(localName, "default");
+
+          const reexport = localData.get(localName);
+          if (reexport) {
```

Comment:
@Kovensky Could you clarify what you mean? `import foo from 'bar'; export { foo }` should be live, but isn't currently, which is a bug: https://github.com/babel/babel/issues/4154 This `if` doesn't handle `export default foo;` though, so that should still be non-live as expected.

Reply:
I'm probably misunderstanding, but I _think_ this changes the internal state to remove the  `export { foo }` and add a metadata equivalent of `export { default as foo } from 'bar'`. If this does what I think it does it fixes 4154, but it would break `export default foo`, since it wouldn't be able to tell `export default foo` from `export { foo as default }` 🤔

Unless the `localData` name is different

Reply:
> it wouldn't be able to tell `export default foo` from `export { foo as default }`

Those two are managed differently. `export default foo`; Isn't part of the `reexport` dataset because it isn't live because it's split up in https://github.com/babel/babel/pull/6230/files#diff-d53cda1dedf7b332fe72eabbf998ac88R269 and treated like

```
var _tmp = foo;
export {_tmp as default};
```
so it would be `_tmp` in the `localData` object, not `foo`.